[red-knot] Port type inference tests to new test framework

Lexxxzy opened the pull request:
## Summary

Porting infer tests to new markdown tests framework.

Link to the corresponding issue: #13696

carljm commented on crates/red_knot_python_semantic/resources/mdtest/imports.md:
`````diff
@@ -0,0 +1,133 @@
+
+We can follow import to class:
+
+````markdown
+```py path=a.py
+from b import C as D; E = D
+reveal_type(E) # revealed: Literal[C]
+```
+
+```py path=b.py
+class C: pass
+```
+````
`````
Oh, the outer markdown fenced code block is just how the mdtest README handles showing an example markdown document within a markdown README, it's not required in the tests and should be removed here and in all other cases:

````suggestion
```py path=a.py
from b import C as D; E = D
reveal_type(E) # revealed: Literal[C]
```

```py path=b.py
class C: pass
```
````

It's funny that it still works either way, because of our naive regex parsing :)

I'll look at updating the README to address this potential confusion.

Lexxxzy replied:
Fixed, my bad. Once again - thanks for clarifying

Lexxxzy commented:
I've changed parser's CODE_RE regex, cause without that fix it was not correctly parsing code snippets with empty files, like:

````markdown
```py path=package/__init__.py
```
````

Of course feel free to comment on this fix...

carljm commented on crates/red_knot_python_semantic/resources/mdtest/imports.md:
`````diff
@@ -0,0 +1,133 @@
+We can follow import to class:
+
+````markdown
+```py path=a.py
`````
I would say it's probably better style in general if we don't specify the path on the "main" test file, if it isn't imported by another file and its location/name don't actually matter to the test. (This wasn't possible in the previous test format.) I don't care too much about this, though, not a blocking issue.

Lexxxzy replied:
Fixed it

carljm reviewed (changes requested):
This is awesome, thank you for taking this on!

I'm going to wait for removal of the extraneous quadruple-backtick markdown fences before reviewing the tests in detail, because removing those will make them a lot easier to read :)

carljm commented:
> I've changed parser's CODE_RE regex

This fix looks good, thank you!

carljm commented:
Also it looks like the mdformat pre-commit check is failing, you can check the ruff CONTRIBUTING doc for info on how to run pre-commit locally so you can catch those issues without having to push and wait for CI.

Lexxxzy commented:
> wait for removal of the extraneous quadruple-backtick

Oh, ok, understood, I will redo that, thanks for clarifing

> Also it looks like the mdformat pre-commit check is failing, you can check the ruff CONTRIBUTING doc for info on how to run pre-commit locally so you can catch those issues without having to push and wait for CI.

Got it, totally forgot about it

Lexxxzy commented:
Rebased on the current `main` branch. Changed format of tests after new clarifications from @carljm. Also added tests for booleans infer.

I have problems with porting two tests so far:
1. `not_literal_string`
In Rust's test there are formatting for variable in tests file like `y = "a".repeat(TypeInferenceBuilder::MAX_STRING_LITERAL_SIZE + 1)`. The question is: how do I correctly insert variable in tests in mdtest the same way?

2. `from_import_with_no_module_name`
````
NOTE: Diagnostic's error is not resolving, also already tried using full message
```py
from import bar   # error: "Expected a module name"
reveal_type(bar)  # revealed: Unknown
```
````

Lexxxzy commented:
Most of the work is done. Now I’m just waiting for feedback on what’s finished and looking for some help with the issues I’m facing before I can keep going.

carljm commented on crates/red_knot_python_semantic/resources/mdtest/conditionals.md:
````diff
@@ -0,0 +1,207 @@
+reveal_type(y)  # revealed: Literal[2, 3, 4]
+```
+
+## Narrowing
+
+### `is not None` check
+
+This test ensures correct type narrowing with an `is not None` check.
````
This test can go in `narrow/not_none.rs`.

And I don't think we need the prose sentence here, it doesn't add anything.

carljm commented on crates/red_knot_python_semantic/resources/mdtest/conditionals.md:
````diff
@@ -0,0 +1,207 @@
+This test ensures correct type narrowing with an `is not None` check.
+
+```py
+x = None if flag else 1
+y = 0
+if x is not None:
+    y = x
+
+reveal_type(x)  # revealed: None | Literal[1]
+reveal_type(y)  # revealed: Literal[0, 1]
````
The existence of `y` in this test was really a hacky workaround for the difficulty of asserting revealed types in locations other than end-of-scope in our previous test style. This is a simpler way to get the same intended test:
```suggestion
x = None if flag else 1
if x is not None:
    reveal_type(x)  # revealed: Literal[1]

reveal_type(x)  # revealed: None | Literal[1]
```

carljm commented on crates/red_knot_python_semantic/resources/mdtest/conditionals.md:
````diff
@@ -0,0 +1,207 @@
+reveal_type(y)  # revealed: Literal[0, 1]
+```
+
+### TODO: Singleton pattern
+
+> TODO: The correct inferred type should be `Literal[0] | None`, but due to simplification limitations, the inferred type is `Literal[0] | None | (Literal[1] & None)`.
+
+```py
+x = None if flag else 1
+y = 0
+match x:
+    case None:
+        y = x
+
+reveal_type(y)  # revealed: Literal[0] | None | Literal[1] & None
+```
````
Let's put this test in `narrow/match.rs`.

And some edits:
````suggestion
### Singleton pattern

```py
x = None if flag else 1
match x:
    case None:
        # TODO intersection simplification: should be just Literal[0] | None
        reveal_type(x)  # revealed: Literal[0] | None | Literal[1] & None
```
````

carljm commented on crates/red_knot_python_semantic/resources/mdtest/exceptions.md:
````diff
@@ -0,0 +1,89 @@
+    reveal_type(f)  # revealed: AttributeError | TypeError
+```
+
+## TODO: Dynamic exception types
+
+> TODO: `e` should be `AttributeError`. `f` should be `OSError | RuntimeError`. `g` should be `BaseException`.
+
+```py
+def foo(x: type[AttributeError], y: tuple[type[OSError], type[RuntimeError]], z: tuple[type[BaseException], ...]):
+    try:
+        w
+    except x as e:
+        reveal_type(e)  # revealed: @Todo
+    except y as f:
+        reveal_type(f)  # revealed: @Todo
+    except z as g:
+        reveal_type(g)  # revealed: @Todo
````
````suggestion
## Dynamic exception types

```py
def foo(x: type[AttributeError], y: tuple[type[OSError], type[RuntimeError]], z: tuple[type[BaseException], ...]):
    try:
        w
    except x as e:
        # TODO should be `AttributeError`
        reveal_type(e)  # revealed: @Todo
    except y as f:
        # TODO should be `OSError | RuntimeError`
        reveal_type(f)  # revealed: @Todo
    except z as g:
        # TODO should be `BaseException`
        reveal_type(g)  # revealed: @Todo
````

carljm commented on crates/red_knot_python_semantic/resources/mdtest/exceptions.md:
```diff
@@ -0,0 +1,89 @@
+
+## TODO: Except star
+
+TODO(Alex): Once we support `sys.version_info` branches, we can set `--target-version=py311` in this tests and the inferred type will just be `BaseExceptionGroup`
```
```suggestion
TODO(Alex): Once we support `sys.version_info` branches, we can set `--target-version=py311` in these tests and the inferred type will just be `BaseExceptionGroup`
```

carljm commented on crates/red_knot_python_semantic/resources/mdtest/functions.md:
````diff
@@ -0,0 +1,176 @@
+reveal_type(x)  # revealed: @Todo
+```
+
+### Constructor
````
This doesn't belong under Functions heading; let's put it in `call/constructor.md`

carljm commented on crates/red_knot_python_semantic/resources/mdtest/functions.md:
```diff
@@ -0,0 +1,176 @@
+# Functions
+
+## Call Expression
```
Let's put these tests in `call/function.rs`

carljm commented on crates/red_knot_python_semantic/resources/mdtest/exceptions.md:
````diff
@@ -0,0 +1,89 @@
+    reveal_type(e)  # revealed: Unknown | BaseExceptionGroup
+```
+
+## TODO: Except\* with specific exception
+
+> TODO(Alex): more precise would be `ExceptionGroup[OSError]`.
+
+```py
+try:
+    x
+except* OSError as e:
+    reveal_type(e)  # revealed: Unknown | BaseExceptionGroup
+```
````
I'll leave the rest of these uncommented to reduce comment volume, but let's follow this kind of pattern everywhere instead:
````suggestion
## Except\* with specific exception

```py
try:
    x
except* OSError as e:
    # TODO(Alex): more precise would be `ExceptionGroup[OSError]`
    reveal_type(e)  # revealed: Unknown | BaseExceptionGroup
```
````

carljm commented on crates/red_knot_python_semantic/resources/mdtest/functions.md:
```diff
@@ -0,0 +1,176 @@
+
+### Parameter
+
+In Python, it's valid to reassign or redeclare a function parameter within the body of the function. In this example, the parameter `x` of type `str` is shadowed and reassigned with a new `int` value inside the function. No diagnostics should be generated.
```
```suggestion
Parameter `x` of type `str` is shadowed and reassigned with a new `int` value inside the function. No diagnostics should be generated.
```

carljm commented on crates/red_knot_python_semantic/resources/mdtest/loops.md:
```diff
@@ -0,0 +1,241 @@
+# Loops
```
Similar to some above cases, let's instead have a `loops/` directory with e.g. `for_loop.md`, `while_loop.md` files in it.

carljm commented on crates/red_knot_python_semantic/resources/mdtest/numbers.md:
````diff
@@ -10,6 +10,15 @@ We can infer an integer literal type:
 ```
 
+### Variable
+
+We can track that variable as an integer:
````
```suggestion
```

carljm commented on crates/red_knot_python_semantic/resources/mdtest/numbers.md:
```diff
@@ -18,6 +27,28 @@ We only track integer literals within the range of an i64:
 
+### Big int
+
+We can track that a big int literal resolves to the corresponding integer literal:
```
```suggestion
We don't support big integer literals; we just infer `int` type instead:
```

carljm commented on crates/red_knot_python_semantic/resources/mdtest/numbers.md:
````diff
@@ -18,6 +27,28 @@ We only track integer literals within the range of an i64:
+```
+
+### Negated
+
+We can track that a negated integer literal resolves to the corresponding integer literal:
````
```suggestion
```

carljm commented on crates/red_knot_python_semantic/resources/mdtest/numbers.md:
````diff
@@ -33,3 +64,79 @@ Same for complex:
 reveal_type(2j)  # revealed: complex
 ```
+
+## Arithmetic
````
I'd rather put these in `unary/integers.md` and `binary/integers.md` files

carljm commented on crates/red_knot_python_semantic/resources/mdtest/strings.md:
````diff
@@ -0,0 +1,129 @@
+reveal_type(z)  # revealed: Literal["No \"closure here"]
+```
+
+## f-strings
````
I'd rather put f-strings in their own file; it can either be a top-level `fstrings.md` or we can add a `literal/` subdir and put several of these files in there.

carljm commented on crates/red_knot_python_semantic/resources/mdtest/strings.md:
````diff
@@ -0,0 +1,129 @@
+reveal_type(a)  # revealed: str
+```
+
+## Subscript
````
These belong in `subscript/string.md`

carljm commented on crates/red_knot_python_semantic/resources/mdtest/strings.md:
````diff
@@ -0,0 +1,129 @@
+reveal_type(a)  # revealed: str
+```
+
+## Bytes
````
Bytes are a separate type from strings, this should go in `bytes.md` or `literal/bytes.md`

carljm commented on crates/red_knot_python_semantic/resources/mdtest/variables.md:
````diff
@@ -0,0 +1,134 @@
+reveal_type(x)  # revealed: Unbound
+```
+
+### Annotation only transparent to local inference
````
All the tests from here onward should go in `shadowing.md` along with some other tests I commented above.

carljm commented on crates/red_knot_python_semantic/resources/mdtest/variables.md:
````diff
@@ -0,0 +1,134 @@
+reveal_type(x)  # revealed: Literal[1, 2]
+```
+
+## Assignment
````
These could go in `assignment.md`

carljm commented on crates/red_knot_python_semantic/resources/mdtest/variables.md:
```diff
@@ -0,0 +1,134 @@
+# Variables
+
+## Union resolution
```
This test can go with some others in `conditional/if_statement.md`

carljm reviewed (commented):
Thank you so much, this is a lot of good work! Excellent translation of the tests to the new assertion style.

A few general comments, which I commented partway through but not everywhere they occur:

- Avoid `TODO` in test titles, and keep TODO comments minimal in length and as close as possible to the specific line/assertion that isn't quite right yet.
- Avoid boilerplate phrases like "This test ensures that" or "Check that" or "We can infer the type when ..." or anything similar -- minimize the words that are not adding specific value to the test in question.
- Break up into more smaller files; I commented on this throughout to try to give a clear idea of an organization that I think would work well.

I appreciate your quick work on this PR and would like to get it landed ASAP to minimize conflicts as new tests are added in other PRs! If you won't have time to update it in the next day-ish, please just let me know, I'm also happy to make the updates and get it landed!

Thanks again.

Lexxxzy commented:
> would like to get it landed ASAP

All clear. Will try to push fixes as soon as possible

Lexxxzy commented:
@carljm 
> A few general comments, which I commented partway through but not everywhere they occur

I tried to follow the style you described. Please take a look at the tests with the new structure. Sorry for all the issues!

Lexxxzy commented:
@sharkdp Thanks for letting me know! I rebased on the current main, but somehow `subscript/string.md - Subscript on strings - Function return` test started to fail. Now it says that revealed type is `@Todo`, and not `str`. Can somebody check what went wrong? Maybe it is because of #5fa82f (Sync vendored typeshed stubs) commit?

Lexxxzy commented:
> Can somebody check what went wrong?

Yep, I checked, actually on @sharkdp pr merge commit (74bf4b) test is running fine, but on 5fa82f it started to fail

AlexWaygood commented:
> > Can somebody check what went wrong?
> 
> Yep, I checked, actually on @sharkdp pr merge commit (74bf4b) test is running fine, but on 5fa82f it started to fail

Yes -- unfortunately I had to change the assertion the test was making as part of that PR (https://github.com/astral-sh/ruff/pull/13753). This is because typeshed now declares `str.__getitem__` to be an overloaded function, and we don't understand overloaded functions yet.

AlexWaygood commented:
This is the specific commit in that PR where I tweaked the test @Lexxxzy: https://github.com/astral-sh/ruff/pull/13753/commits/8ec3b30ea480cd63c01c3ef54aed73517c343196

Sorry for the bother!

Lexxxzy commented:
> You'll need to change the assertion to @Todo in your mdtest port of the test as well.

Ok, i see. Thanks for the info!

carljm reviewed (approved):
This is great!! Thank you so much for the quick and thorough work. I'm going to make a few minor cosmetic tweaks, push those, wait for CI, and then merge.

AlexWaygood commented:
Thanks so much for taking this on @Lexxxzy! A big first contribution, and a really helpful one for us!!
